### ISSUE TITLE
Telegram is incompatible with VoiceOver

### ISSUE THREAD

**Issue by KreerC** [peter-iakovlev/Telegram#123]
All non-native iOS components of the Telegram app, mainly the chat window, are incompatible with VoiceOver, and so unusable for the blind and visually impaired who are dependent on this accessibility feature. This should receive a fix ASAP.

**Comment by Nayomi1202** [comment]
Thank-you nkreer for bringing this up! I have tried contacting Telegram and they don't see to mind at all. Do you have any updates?

**Comment by KreerC** [comment]
The new version 4.1 of the app seems to have made it even worse. Guess I won't be a Telegram user for long anymore.

**Comment by Nayomi1202** [comment]
I know it's annoying how ignorant they are toward the visually impaired. I am with you Peter!

**Comment by JulioBBL** [comment]
i'm learning to implement voice over and would like to help, but the app is written in objective-c, i know swift, looks like we will have to wait.

**Comment by utro5750** [comment]
Telegram X was written on Swift. Our turn?

**Comment by LiorElisha** [comment]
A good friend is visually impaired, he has iPhone 8 and it's still doesn't work! 
I'm shocked that a big company like Telegram is not accessible.
He cannot use the app.

Help!!

**Comment by hedw1gP** [comment]
Telegram 5.0 (basically X for release) is out and although it is Swift, it still cannot use VoiceOver

**Comment by dpy013** [comment]
Hi I am a blind person from China. When I was preparing to use Telegram, I found that voice over could not operate him at all. This surprised me. I have been exposed to foreign software for so many years. Telegram is the first time I discovered it. Voice over An app that can't be operated at all.

**Comment by pslacerda** [comment]
The Android app also don't handle VoiceOver. If I had time and knowledge I would fix it by myself.

Please fix it.

https://github.com/peter-iakovlev/TelegramCore/issues/46

> Hi I am a blind person from China. When I was preparing to use Telegram, I found that voice over could not operate him at all. This surprised me. I have been exposed to foreign software for so many years. Telegram is the first time I discovered it. Voice over An app that can't be operated at all.

I just installed IM+ Messenger for enable a friend of mine to use Telegram on her OS X, it simply open a Chrome frame for Telegram web and she didn't liked it. The native one is way better.

*Update*

Related documentation: https://developer.apple.com/documentation/uikit/accessibility/uiaccessibility#//apple_ref/doc/constant_group/Accessibility_Traits

**Comment by dw-b** [comment]
There's still no support in either Telegram Messenger for iOS (5.1.1) or Telegram X (5.0.17) for Apple VoiceOver: both apps are totally inaccessible to visually impaired people.

Other mainstream messengers have provided accessibility support for years and it's not particularly difficult to add basic support (e.g. accessible labels on buttons).

(https://developer.apple.com/documentation/uikit/accessibility/supporting_voiceover_in_your_app)

**Comment by pslacerda** [comment]
@dw-b

The widgets are in another Objective-C project (https://github.com/peter-iakovlev/Display).

Someone could be assigned to give such accessibility traits on widgets, my friend is blind and we can test your developments.
